lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp

```diff
@@ -435,6 +435,45 @@ static LogicalResult applyOutputDirAnno(const AnnoPathValue &target,
   return success();
 }
 
+/// Convert from FullAsyncResetAnnotation to FullResetAnnotation
+static LogicalResult convertToFullResetAnnotation(const AnnoPathValue &target,
+                                                  DictionaryAttr anno,
+                                                  ApplyState &state) {
+  auto *op = target.ref.getOp();
+  auto *context = op->getContext();
+
+  mlir::emitWarning(op->getLoc())
+      << "'" << fullAsyncResetAnnoClass << "' is deprecated, use '"
+      << fullResetAnnoClass << "' instead";
+
+  NamedAttrList newAnno(anno.getValue());
+  newAnno.set("class", StringAttr::get(context, fullResetAnnoClass));
+  newAnno.append("resetType", StringAttr::get(context, "async"));
+
+  DictionaryAttr newDictionary = DictionaryAttr::get(op->getContext(), newAnno);
+
+  return applyWithoutTarget<false>(target, newDictionary, state);
+}
+
+/// Convert from IgnoreFullAsyncResetAnnotation to
+/// ExcludeFromFullResetAnnotation
+static LogicalResult convertToExcludeFromFullResetAnnotation(
+    const AnnoPathValue &target, DictionaryAttr anno, ApplyState &state) {
+  auto *op = target.ref.getOp();
+  auto *context = op->getContext();
+
+  mlir::emitWarning(op->getLoc())
+      << "'" << ignoreFullAsyncResetAnnoClass << "' is deprecated, use '"
+      << excludeFromFullResetAnnoClass << "' instead";
+
+  NamedAttrList newAnno(anno.getValue());
+  newAnno.set("class", StringAttr::get(context, excludeFromFullResetAnnoClass));
+
+  DictionaryAttr newDictionary = DictionaryAttr::get(op->getContext(), newAnno);
+
+  return applyWithoutTarget<true, FModuleOp>(target, newDictionary, state);
+}
```

Comment:
You can probably avoid this much code reuse with a templated fucntion or simply one that takes the specifics as parameters and then does the annotation update separately.

Reply:
I'm not sure my C++ is quite at the level of templating a function sufficiently to deal with it needing to call differently templated functions (see 455 and 475), maybe that could be handled by accepting a thunk. It'd be easy for me in Scala 😂.